### PR TITLE
Fix mass update or delete lost records, closes #320

### LINKED ISSUE #320: DuplicateKey after delete all then insert
In 0.8.0 when executing a full table delete then an insert sometimes happens that insert operation generate a DuplicateKeyException when no data should exist in the table itself (after delete)

```
Caused by: java.sql.SQLException: java.util.concurrent.CompletionException: herddb.model.DuplicatePrimaryKeyException: key 0400000001040000001f, decoded as {xxx=31, yyy=1}, already exists in table kkk during transaction 1433
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:870)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
        at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2251)
        at herddb.core.TableManager.executeStatementAsync(TableManager.java:530)
        at herddb.core.TableSpaceManager.executeTableAwareStatement(TableSpaceManager.java:1169)
        at herddb.core.TableSpaceManager.executeStatementAsyncInternal(TableSpaceManager.java:1073)
        at herddb.core.TableSpaceManager.executeStatementAsync(TableSpaceManager.java:1053)
        at herddb.core.DBManager.executeStatementAsync(DBManager.java:634)
        at herddb.core.DBManager.executePlanAsync(DBManager.java:691)
        at herddb.server.ServerSideConnectionPeer$1ComputeNext.accept(ServerSideConnectionPeer.java:661)
        at herddb.server.ServerSideConnectionPeer$1ComputeNext.accept(ServerSideConnectionPeer.java:598)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:610)
        at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649)
        at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: herddb.model.DuplicatePrimaryKeyException: key 0400000001040000001f, decoded as {xxx=31, yyy=1}, already exists in table kkk during transaction 1433
        at herddb.core.TableManager.executeInsertAsync(TableManager.java:899)
        at herddb.core.TableManager.executeStatementAsync(TableManager.java:512)
        ... 16 more

        at herddb.jdbc.utils.SQLExceptionUtils.wrapException(SQLExceptionUtils.java:40)
        at herddb.jdbc.HerdDBPreparedStatement.executeBatch(HerdDBPreparedStatement.java:242)
        at magnews.gateway.database.Database.lambda$updateIpProviderStatusData$8(Database.java:654)
        at magnews.gateway.database.TransactionHelper$DatabaseWork.lambda$fromVoid$0(TransactionHelper.java:37)
        at magnews.gateway.database.TransactionHelper.doInTransaction(TransactionHelper.java:334)
        ... 8 more
Caused by: herddb.client.HDBException: java.util.concurrent.CompletionException: herddb.model.DuplicatePrimaryKeyException: key 0400000001040000001f, decoded as {xxx=31, yyy=1}, already exists in table kkk during transaction 1433
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:870)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
        at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2251)
        at herddb.core.TableManager.executeStatementAsync(TableManager.java:530)
        at herddb.core.TableSpaceManager.executeTableAwareStatement(TableSpaceManager.java:1169)
        at herddb.core.TableSpaceManager.executeStatementAsyncInternal(TableSpaceManager.java:1073)
        at herddb.core.TableSpaceManager.executeStatementAsync(TableSpaceManager.java:1053)
        at herddb.core.DBManager.executeStatementAsync(DBManager.java:634)
        at herddb.core.DBManager.executePlanAsync(DBManager.java:691)
        at herddb.server.ServerSideConnectionPeer$1ComputeNext.accept(ServerSideConnectionPeer.java:661)
        at herddb.server.ServerSideConnectionPeer$1ComputeNext.accept(ServerSideConnectionPeer.java:598)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:610)
        at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649)
        at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: herddb.model.DuplicatePrimaryKeyException: key 0400000001040000001f, decoded as {xxx=31, yyy=1}, already exists in table kkk during transaction 1433
        at herddb.core.TableManager.executeInsertAsync(TableManager.java:899)
        at herddb.core.TableManager.executeStatementAsync(TableManager.java:512)
        ... 16 more

        at herddb.client.RoutedClientSideConnection.handleGenericError(RoutedClientSideConnection.java:624)
        at herddb.client.RoutedClientSideConnection.handleGenericError(RoutedClientSideConnection.java:606)
        at herddb.client.RoutedClientSideConnection.executeUpdates(RoutedClientSideConnection.java:408)
        at herddb.client.HDBConnection.executeUpdates(HDBConnection.java:274)
        at herddb.jdbc.HerdDBPreparedStatement.executeBatch(HerdDBPreparedStatement.java:230)
        ... 11 more
```

### PR DESCRIPTION
**Pull request by diegosalvi**
During mass deletes or updates some records were sometimes "lost" in transaction and no modified at all (due to a concurrent "unaligned" commit position evaluation)

### CODE REVIEW

File: herddb-core/src/main/java/herddb/core/TableManager.java [hunk 1]
```diff
@@ -1007,44 +1009,75 @@ public void accept(Record actual, LockHandle lockHandle) throws StatementExecuti
...
+                    }, tableSpaceManager.getCallbacksExecutor())
+                    .thenApply((pendings) -> {
                         return new DMLStatementExecutionResult(transactionId, updateCount.get(), lastKey.value,
                                 update.isReturnValues() ? (lastValue.value != null ? Bytes.from_array(lastValue.value) : null) : null);
                     });
         }
 
     }
 
+    private final class PendingLogEntryWork {
```
[the remark this answers is not in the document]
Comment (diegosalvi) [reply herddb-core/src/main/java/herddb/core/TableManager.java]: Right! Good catch

File: herddb-jdbc/src/test/java/herddb/jdbc/UpdateThenSelectTest.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+import herddb.server.Server;
+import herddb.server.ServerConfiguration;
+import herddb.server.StaticClientSideMetadataProvider;
+
+/**
+ * Check update then select (#320, not every row update really was run).
+ *
+ * @author diego.salvi
+ */
+public class UpdateThenSelectTest {
```
[the remark this answers is not in the document]
Comment (diegosalvi) [reply herddb-jdbc/src/test/java/herddb/jdbc/UpdateThenSelectTest.java]: Dropped and added similar into core. No need of duplicated test cases

File: herddb-core/src/main/java/herddb/core/TableManager.java [hunk 3]
```diff
@@ -1053,43 +1086,62 @@ public void accept(Record actual, LockHandle lockHandle) throws StatementExecuti
...
-                return new DMLStatementExecutionResult(transactionId, updateCount.get(), lastKey.value,
-                        delete.isReturnValues() ? (lastValue.value != null ? Bytes.from_array(lastValue.value) : null) : null);
-            });
+        }
+
+        if (writes.size() == 1) {
+            return writes.get(0)
+                    .whenCompleteAsync((pending, error) -> {
+                        try {
+                            if (error == null) {
```
[the remark this answers is not in the document]
Comment (diegosalvi) [reply herddb-core/src/main/java/herddb/core/TableManager.java]: Won't already be in error?

From javadoc:

> Returns a new CompletionStage with the same result or exception as this stage, that executes the given action using this stage's default asynchronous execution facility when this stage completes.

I had to test it but it seems to me working like that

[the remark this answers is not in the document]
Comment (eolivelli) [reply herddb-core/src/main/java/herddb/core/TableManager.java]: okay

**Pull request merged**